src/algebra/direct_limit.lean

```diff
@@ -0,0 +1,483 @@
+/-
+Copyright (c) 2019 Kenny Lau, Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Chris Hughes
+
+Direct limit of modules, abelian groups, rings, and fields.
+-/
+
+import linear_algebra.direct_sum_module
+import linear_algebra.linear_combination
+import ring_theory.free_comm_ring
+import ring_theory.ideal_operations
+
+universes u v w u₁
+
+open lattice submodule
+
+variables {R : Type u} [ring R]
+variables {ι : Type v} [nonempty ι]
+variables [directed_order ι] [decidable_eq ι]
+variables (G : ι → Type w) [Π i, decidable_eq (G i)]
+
+class directed_system (f : Π i j, i ≤ j → G i → G j) : Prop :=
+(Hid : ∀ i x h, f i i h x = x)
+(Hcomp : ∀ i j k hij hjk x, f j k hjk (f i j hij x) = f i k (le_trans hij hjk) x)
+
+namespace module
+
+variables [Π i, add_comm_group (G i)] [Π i, module R (G i)]
+
+class directed_system (f : Π i j, i ≤ j → G i →ₗ[R] G j) : Prop :=
+(Hid : ∀ i x h, f i i h x = x)
+(Hcomp : ∀ i j k hij hjk x, f j k hjk (f i j hij x) = f i k (le_trans hij hjk) x)
+
```

Comment:
`Hid` and `Hcomp` are not great names here.

Reply:
changed to `map_self` and `map_map`